test/api_spec.js

```diff
@@ -13,6 +13,7 @@ describe("API Tests", function () {
   var apiPort = port + 1;
   var proxy;
   var apiUrl = "http://127.0.0.1:" + apiPort + "/api/routes";
+  var metricsUrl = "http://127.0.0.1:" + apiPort + "/metrics";
```

Comment:
Will the /metrics endpoint be exposed like any other on the same port? If not, what decides if it listens to localhost only or any incoming requests from other IPs etc?

Do you have a suggestion on how we handle this /metrics endpoint from a security standpoint, where I assume we prefer to not expose it publicly or preferably at least be able to control that somehow.

Reply:
From my perspective, the metrics endpoint should not be public to the internet. However, putting it on the API server together may cause some problems if the API server is protected with API key or/and a client TLS certification, which is not available in your Prometheus server. On the other hand, if you don't use API key nor a client TLS certification, having another server may be too much. So, ideally speaking, we should have 2 options: add the metrics endpoint in the API server, or create another dedicated server for the metrics endpoint.

What do you think?

Reply:
Ah hmmm. I think I'm positive towards exposing it via the API server on /metrics, but not having that route require a API token. Or creating a dedicated server for the metrics endpoint...

Not confident about this, hmmm... I know i dislike re-using an API server access token for access to metrics endpoint at least.

Reply:
Perhaps the most flexible option for the future is to have a dedicated server? I dont know how advanced it is to go en for the various options, and reducing technical complexity is also valuable.

Reply:
> Perhaps the most flexible option for the future is to have a dedicated server? 

Agree. I think it's the most flexible. We can drop the metrics endpoint in the API server entirely because it's complicated and may confuse users if we have 2 options to serve metrics.

Reply:
@dtaniwaki and @minrk, what do you think?

1. Should we expose the metrics on: the proxy server (typically proxying traffic), the proxy-api server (typically controlling traffic routing), and/or a dedicated metrics server (just serving metrics).
2. What if any access control is implemented?
   1. A fixed access token set via an environment variable?
   2. A way to limit what network traffic it will accept traffic _from_, does it accept traffic only from localhost, some local network, or all of internet etc?
3. What should we aim for in this PR, and what should we aim for in future work?

Currently I think:
1. Only a dedicated metrics server
2.
   1. A bonus to have this
   2. It feels almost required we have this
3. Aim for in this PR: a dedicated server that listens to a network interface that can accept traffic from a local network. Future improvements: a fixed access token set via an environment variable.

Reply:
I'm to a large extent thinking about the following situations:
- This software is running on a VM where JupyterHub and Prometheus is also running
- This software is running in a k8s Pod in a k8s cluster, and prometheus is running in another pod in the k8s cluster.

I care about the following:
1. To not expose /metrics to everyone that can be routed
2. To be able to expose /metrics to prometheus if it is running either on the same machine or in the same k8s cluster
3. To protect /metrics with some password dedicated to this purpose even though we don't have TLS communication

More ambitious than I'm interested in:
4. To protect the /metrics endpoint with some password over a TLS/HTTPS connection using a self-signed cert or provided cert

Reply:
I think running it under it's own configurable port may be the easiest for now. You don't have to worry about authentication, a JupyterHub admin shouldn't have to worry too much about the new port if they're following best practice and are running a firewall, and it can be easily blocked on k8s by a network policy.

Reply:
👍 to its own `--metrics-port` (and `--metrics-ip`) config for its own server, and disabled if unspecified (default).